### PR TITLE
Facilities to move derivatives between public and restricted storage type

### PR DESCRIPTION
**Pull request by jrochkind**
Ref #832

We catch in ActiveRecord callback if derivative storage type has changed, and execute a background job to move derivatives between storage locations if so. 

* Bg job also makes sure to delete any DZI files for `restricted` storage; we have no way to serve DZI's but from public storage.  It does *not* create DZI files if you move a file from `restricted` to `public`. Images aren't really our use case at all anyway. 

* Not currently handling removing things from derivatives backup bucket, leave that for another ticket. (Is a checkbox in master Issue, to help us not  forget)

### CODE REVIEW

File: app/models/asset.rb [hunk 1]
```diff
@@ -50,4 +70,12 @@ def self.all_derivative_count
       "select count(*) from (SELECT id, jsonb_object_keys(file_data -> 'derivatives') FROM kithe_models WHERE kithe_model_type = 2) AS asset_derivative_keys;"
     ).first["count"]
   end
+
+  # If derivative_storage_type changed in last save, fire off bg job
+  # to move derivatives to correct place.
+  def ensure_correct_derivatives_storage_after_change
```
Comment (eddierubeiz) [inline app/models/asset.rb:76]: I like this declarative approach.

**Comment by jrochkind** [comment]
@eddierubeiz So I used a callback in the model to automatically move derivatives whenever the attribute is changed. By launching a bg job. 

I wonder if that's too clever, and I should remove it from the model, and just have the controller do it upon form submission from the form on the admin asset show page that lets you change `derivative_storage_type`?  And I wonder if the form on on the admin asset show page should have a way to "copy derivatives to correct location" without changing `derivative_storage_type`, in cases where they may be mis-matched?

**Comment by eddierubeiz** [comment]
_So I used a callback in the model to automatically move derivatives whenever the attribute is changed. By launching a bg job.

I wonder if that's too clever, and I should remove it from the model, and just have the controller do it upon form submission from the form on the admin asset show page that lets you change derivative_storage_type?_

I like having this on the model. I'm thinking of a case in which we might want to invoke the "move derivatives" code on a large number of models from a rake task.

_And I wonder if the form on on the admin asset show page should have a way to "copy derivatives to correct location" without changing derivative_storage_type, in cases where they may be mis-matched?_

I'm not opposed to this idea, but it might be unnecessary. I'd consider not including it in this PR; rather, you could create a someday/maybe issue that we can delete if we find the feature isn't needed in practice.

**Comment by jrochkind** [comment]
Cool thanks for feedback and merging. 

Some people don't like this callback approach because it makes it infeasible to *turn it off* if you want to turn it off -- like maybe you want to change the attribute on hundreds of  Assets *without* triggering hundreds of bg jobs. As a model callback makes it pretty hard to disable when using standard ActiveRecord functions. 

Or alternately, if you wind up changing the value in the db *without* using standard AR functions, it wont' trigger the callbacks, but you might forget you need to move things over yourself anyway because of being used to not thinking about it as "automatic". 

But we could pick through details on this forever, appreciate the feedback, and the decision to just go forward with it as is for now!

**Comment by jrochkind** [comment]
@eddierubeiz Also when quoting someone in a github comment using markdown, you can use `>` markdown for a quote, instead of trying to use `_` for italics. 

> This is a quote on a line that begins with `>`

A bit more readable.

**Pull request merged**